### DISCUSSION TITLE
Vsync via tty

### DISCUSSION THREAD

**Discussion by neurocyte** [kovidgoyal/kitty#6590] [General]
Hi,

is there some way of syncing tty output to the vsync of the monitor that kitty is displaying on?

I am working on a full screen terminal application (an editor) that can display a potentially arbitrary number of frames for various animations. I would like the animations to be smooth, but not waste resources unnecessarily. Ideally I would like some sort of vsync event from the terminal. Or if that is not possible, some way to query the maximum usable frame rate to use as a limit inside my application.

Any thoughts or advice or links to apps that do something similar? Is this perhaps a candidate for a kitty terminal protocol extension?

**Comment by kovidgoyal** [comment]
No, there is no such facility. Use the synchronized screen update escape
code CSI 2026 to ensure your screen updates are atomic. I suggest
drawing at 60 FPS that will give you smooth animations in almost all
cases.

**Reply by neurocyte** [reply]
Yes, I already use CSI 2026 and have my framerate limited to 60fps as that is pretty much supported everywhere. All of my displays are at least 120Hz though (or much higher) and I would like to dynamically take advantage of that. At the moment it takes manual configuration which I would like to avoid.

**Comment by kovidgoyal** [comment]
Well there's no existing protocol for this, and for instance on wayland,
rendering can be rate limited by the compositor using render frames. On
macOS its a constant rate clock. No clue what it is on X11, but kitty
itself defaults to 100FPS. See the repaint_delay and sync_to_monitor
options in kitty.conf

I'll be happy to accept a PR for this, you can simply add another
parameter to the kitty query terminal protocol.
https://sw.kovidgoyal.net/kitty/kittens/query_terminal/

Call it render rate and have kitty report the effective rate taking into
account sync_to_monitor and repaint_delay and the rate of the monitor
the window is displayed on.

There is also the question of what happens if the window is moved to a
new monitor.

**Reply by neurocyte** [reply]
Thanks for your response. I will consider creating a PR when I find the time.
